model/src/main/proto/oppia_logger.proto

```diff
@@ -80,6 +110,105 @@ message EventLog {
+    ExplorationData exploration_data= 2;
+    // The linked skill id of the exploration.
+    string skill_id = 3;
+  }
+
+  // Structure of a end card context.
+  message EndCardContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+    // The linked skill id of the exploration.
+    string skill_id = 3;
+  }
+
+  // Structure of a hint offered context.
+  message HintOfferedContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+    // index of the last revealed hint within the hint list of a state (meaning the index
+    // corresponding to this value is the latest hint that's been revealed by the user)
+    string hint_index = 3;
+  }
+
+  // Structure of an access hint context.
+  message AccessHintContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+    // index of the last revealed hint within the hint list of a state (meaning the index
+    // corresponding to this value is the latest hint that's been revealed by the user)
+    string hint_index = 3;
+  }
+
+  // Structure of an access solution context.
+  message AccessSolutionContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+  }
+
+  // Structure of a solution offered context.
+  message SolutionOfferedContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+  }
+
+  // Structure of a submit answer context.
+  message SubmitAnswerContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+    // Whether the answer that the learner submitted is the correct answer.
+    bool is_answer_correct = 3;
+  }
+
+  // Structure of a play voiceover context.
+  message PlayVoiceOverContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+    // The content id of the voiceover recording.
+    string content_id = 3;
+  }
+
+  // Structure of an app in background context.
+  message AppInBackgroundContext {
+    GenericData generic_data = 1;
+  }
+
+  // Structure of an app in foreground context.
+  message AppInForegroundContext {
+    GenericData generic_data = 1;
+  }
+
+  // Structure of an exit exploration context.
+  message ExitExplorationContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+  }
+
+  // Structure of a finish exploration context.
+  message FinishExplorationContext {
+    GenericData generic_data = 1;
+    ExplorationData exploration_data= 2;
+  }
+
+  // Structure of a resume exploration context.
+  message ResumeExplorationContext {
+    GenericData generic_data = 1;
+  }
+
+  // Structure of a start over exploration context.
+  message StartOverExplorationContext {
+    GenericData generic_data = 1;
+  }
+
+  // Structure of a delete profile context.
+  message DeleteProfileContext {
+    GenericData generic_data = 1;
+  }
+
   // Supported event actions for event logging.
   enum EventAction {
```

Comment:
I'm not necessarily suggesting we change a lot in the analytics controller, but why can't we just use the generated oneof index for these instead of having a separate enum if the two lists are being kept in parity?

Reply:
The idea behind keeping both of them was to indicate what we intend to do (action) and what we intend to collect (context data) in a mutually exclusive manner. We might come across a situation where we collect similar data across different event actions. For example: event actions like `OPEN_INFO_TAB`, `OPEN_LESSONS_TAB`, `OPEN_PRACTICE_TAB` and `OPEN_REVISION_TAB`, correspond to a single context - `TopicContext`. Apart from this, the values of this enum (EventAction) are also the event names that we send to Firebase. Does that help?

Reply:
I don't think that quite addresses why a oneof wouldn't work. In your example, each of those actions would presumably be different oneof cases with the same type (i.e. distinct events with the same logged details), e.g.:

```kotlin
TopicContext open_info_tab = 1;
TopicContext open_lessons_tab = 2;
TopicContext open_practice_tab = 3;
TopicContext open_revision_tab = 4;
...
```

Or, am I missing something?

Reply:
Removed event action and modified dependencies to depend directly on the context. One thing while implementing that I was kind of confused about was handling of cases where we don't log any context -- so I've created a null message for these type of cases.

Reply:
Left a follow up comment separately for the missing context case, otherwise I think the changes look really good.